TE9 [TC-DD-3.20] `chip-tool-darwin` Support Needed: Unpairing command

#### Problem
Support is needed for unpairing `chip-tool-darwin` from a TH=Accessory for Step 3 in `TC-DD-3.20`

![image](https://user-images.githubusercontent.com/80923273/167705342-77c0c59d-5cb5-4f22-a548-376c15da5b9a.png)

 
```
$ ./out/chip-tool-darwin/chip-tool-darwin pairing unpair 1
[1652137419748] [71098:6216094] CHIP: [TOO] Unknown command: unpair
Usage:
  ./out/chip-tool-darwin/chip-tool-darwin pairing command_name [param1 param2 ...]

  +-------------------------------------------------------------------------------------+
  | Commands:                                                                           |
  +-------------------------------------------------------------------------------------+
  | * qrcode                                                                            |
  | * manualcode                                                                        |
  | * ethernet                                                                          |
  | * ble-wifi                                                                          |
  | * ble-thread                                                                        |
  +-------------------------------------------------------------------------------------+
[1652137419749] [71098:6216094] CHIP: [TOO] Run command failure: ../../examples/chip-tool-darwin/third_party/connectedhomeip/examples/chip-tool/commands/common/Commands.cpp:96: Error 0x0000002F
```

#### Test Cases
[3.3.20. [TC-DD-3.20] Commissioning Flow - Commission, Unpair and Re-commission Device [DUT - Commissioner]
](https://github.com/CHIP-Specifications/chip-test-plans/blob/master/src/devicediscovery.adoc#3320-tc-dd-320-commissioning-flow---commission-unpair-and-re-commission-device-dut---commissioner)

--- a/examples/chip-tool-darwin/commands/pairing/Commands.h
+++ b/examples/chip-tool-darwin/commands/pairing/Commands.h
@@ -50,13 +50,19 @@ public:
     PairBleThread() : PairingCommandBridge("ble-thread", PairingMode::Ble, PairingNetworkType::Thread) {}
 };
 
+class Unpair : public PairingCommandBridge
+{
+public:
+    Unpair() : PairingCommandBridge("unpair", PairingMode::None, PairingNetworkType::None) {}
+};
+
 void registerCommandsPairing(Commands & commands)
 {
     const char * clusterName = "Pairing";
 
     commands_list clusterCommands = {
         make_unique<PairQRCode>(),  make_unique<PairManualCode>(), make_unique<PairWithIPAddress>(),
-        make_unique<PairBleWiFi>(), make_unique<PairBleThread>(),
+        make_unique<PairBleWiFi>(), make_unique<PairBleThread>(),  make_unique<Unpair>(),
     };
 
     commands.Register(clusterName, clusterCommands);

--- a/examples/chip-tool-darwin/commands/pairing/PairingCommandBridge.h
+++ b/examples/chip-tool-darwin/commands/pairing/PairingCommandBridge.h
@@ -22,6 +22,7 @@
 
 enum class PairingMode
 {
+    None,
     QRCode,
     ManualCode,
     Ethernet,
@@ -59,6 +60,8 @@ public:
 
         switch (mode)
         {
+        case PairingMode::None:
+            break;
         case PairingMode::QRCode:
             AddArgument("payload", &mOnboardingPayload);
             break;
@@ -86,6 +89,7 @@ private:
     void PairWithCode(NSError * __autoreleasing * error);
     void PairWithPayload(NSError * __autoreleasing * error);
     void PairWithIPAddress(NSError * __autoreleasing * error);
+    void Unpair();
     void SetUpPairingDelegate();
 
     const PairingMode mPairingMode;

--- a/examples/chip-tool-darwin/commands/pairing/PairingCommandBridge.mm
+++ b/examples/chip-tool-darwin/commands/pairing/PairingCommandBridge.mm
@@ -17,6 +17,7 @@
  */
 
 #import <CHIP/CHIP.h>
+#import <CHIP/CHIPError_Internal.h>
 
 #include "../common/CHIPCommandBridge.h"
 #include "PairingCommandBridge.h"
@@ -57,6 +58,9 @@ CHIP_ERROR PairingCommandBridge::RunCommand()
 {
     NSError * error;
     switch (mPairingMode) {
+    case PairingMode::None:
+        Unpair();
+        break;
     case PairingMode::QRCode:
     case PairingMode::ManualCode:
         PairWithPayload(&error);
@@ -97,4 +101,52 @@ void PairingCommandBridge::PairWithIPAddress(NSError * __autoreleasing * error)
                                  port:mRemotePort
                          setupPINCode:mSetupPINCode
                                 error:error];
+}
+
+void PairingCommandBridge::Unpair()
+{
+    dispatch_queue_t callbackQueue = dispatch_queue_create("com.chip-tool.command", DISPATCH_QUEUE_SERIAL);
+    [CurrentCommissioner()
+        getConnectedDevice:mNodeId
+                     queue:callbackQueue
+         completionHandler:^(CHIPDevice * _Nullable device, NSError * _Nullable error) {
+             CHIP_ERROR err = CHIP_NO_ERROR;
+             if (error) {
+                 err = [CHIPError errorToCHIPErrorCode:error];
+                 LogNSError("Error: ", error);
+                 SetCommandExitStatus(err);
+             } else if (device == nil) {
+                 ChipLogError(chipTool, "Error: %s", chip::ErrorStr(CHIP_ERROR_INTERNAL));
+                 SetCommandExitStatus(CHIP_ERROR_INTERNAL);
+             } else {
+                 ChipLogProgress(chipTool, "Attempting to unpair device %llu", mNodeId);
+                 CHIPOperationalCredentials * opCredsCluster = [[CHIPOperationalCredentials alloc] initWithDevice:device
+                                                                                                         endpoint:0
+                                                                                                            queue:callbackQueue];
+                 [opCredsCluster readAttributeCurrentFabricIndexWithCompletionHandler:^(
+                     NSNumber * _Nullable value, NSError * _Nullable readError) {
+                     if (readError) {
+                         CHIP_ERROR readErr = [CHIPError errorToCHIPErrorCode:readError];
+                         LogNSError("Failed to get current fabric: ", readError);
+                         SetCommandExitStatus(readErr);
+                         return;
+                     }
+                     CHIPOperationalCredentialsClusterRemoveFabricParams * params =
+                         [[CHIPOperationalCredentialsClusterRemoveFabricParams alloc] init];
+                     params.fabricIndex = value;
+                     [opCredsCluster removeFabricWithParams:params
+                                          completionHandler:^(CHIPOperationalCredentialsClusterNOCResponseParams * _Nullable data,
+                                              NSError * _Nullable removeError) {
+                                              CHIP_ERROR removeErr = CHIP_NO_ERROR;
+                                              if (removeError) {
+                                                  removeErr = [CHIPError errorToCHIPErrorCode:removeError];
+                                                  LogNSError("Failed to remove current fabric: ", removeError);
+                                              } else {
+                                                  ChipLogProgress(chipTool, "Successfully unpaired deviceId %llu", mNodeId);
+                                              }
+                                              SetCommandExitStatus(removeErr);
+                                          }];
+                 }];
+             }
+         }];
 }

